src/magneticField/turbulentField/GridTurbulence.cpp

```diff
@@ -0,0 +1,198 @@
+				phase = 2 * M_PI * random.rand();
+				cosPhase = cos(phase); // real part
+				sinPhase = sin(phase); // imaginary part
+
+				Bkx[i][0] = b.x * cosPhase;
+				Bkx[i][1] = b.x * sinPhase;
+				Bky[i][0] = b.y * cosPhase;
+				Bky[i][1] = b.y * sinPhase;
+				Bkz[i][0] = b.z * cosPhase;
+				Bkz[i][1] = b.z * sinPhase;
+			} // for iz
+		}     // for iy
+	}         // for ix
+
+	executeInverseFFTInplace(gridPtr, Bkx, Bky, Bkz);
+
+	fftwf_free(Bkx);
+	fftwf_free(Bky);
+	fftwf_free(Bkz);
+
+	scaleGrid(gridPtr, spectrum.getBrms() /
+	                       rmsFieldStrength(gridPtr)); // normalize to Brms
+}
+
+// Check the grid properties before the FFT procedure
+void GridTurbulence::checkGridRequirements(ref_ptr<Grid3f> grid, double lMin,
+                                           double lMax) {
+	size_t Nx = grid->getNx();
+	size_t Ny = grid->getNy();
+	size_t Nz = grid->getNz();
+	Vector3d spacing = grid->getSpacing();
+
+	if ((Nx != Ny) or (Ny != Nz))
+		throw std::runtime_error("turbulentField: only cubic grid supported");
+	if ((spacing.x != spacing.y) or (spacing.y != spacing.z))
+		throw std::runtime_error("turbulentField: only equal spacing suported");
+	if (lMin < 2 * spacing.x)
+		throw std::runtime_error("turbulentField: lMin < 2 * spacing");
+	if (lMin >= lMax)
+		throw std::runtime_error("turbulentField: lMin >= lMax");
+	if (lMax > Nx * spacing.x) // before was (lMax > Nx * spacing.x / 2)
+		throw std::runtime_error("turbulentField: lMax > size");
+}
+
+// Execute inverse discrete FFT in-place for a 3D grid, from complex to real
+// space
+void GridTurbulence::executeInverseFFTInplace(ref_ptr<Grid3f> grid,
+                                              fftwf_complex *Bkx,
+                                              fftwf_complex *Bky,
+                                              fftwf_complex *Bkz) {
+
+	size_t n = grid->getNx(); // size of array
+	size_t n2 = (size_t)floor(n / 2) +
+	            1; // size array in z-direction in configuration space
+
+	// in-place, complex to real, inverse Fourier transformation on each
+	// component note that the last elements of B(x) are unused now
+	float *Bx = (float *)Bkx;
+	fftwf_plan plan_x = fftwf_plan_dft_c2r_3d(n, n, n, Bkx, Bx, FFTW_ESTIMATE);
+	fftwf_execute(plan_x);
+	fftwf_destroy_plan(plan_x);
+
+	float *By = (float *)Bky;
+	fftwf_plan plan_y = fftwf_plan_dft_c2r_3d(n, n, n, Bky, By, FFTW_ESTIMATE);
+	fftwf_execute(plan_y);
+	fftwf_destroy_plan(plan_y);
+
+	float *Bz = (float *)Bkz;
+	fftwf_plan plan_z = fftwf_plan_dft_c2r_3d(n, n, n, Bkz, Bz, FFTW_ESTIMATE);
+	fftwf_execute(plan_z);
+	fftwf_destroy_plan(plan_z);
+
+	// save to grid
+	int i;
+	for (size_t ix = 0; ix < n; ix++) {
+		for (size_t iy = 0; iy < n; iy++) {
+			for (size_t iz = 0; iz < n; iz++) {
+				i = ix * n * 2 * n2 + iy * 2 * n2 + iz;
+				Vector3f &b = grid->get(ix, iy, iz);
+				b.x = Bx[i];
+				b.y = By[i];
+				b.z = Bz[i];
+			}
+		}
+	}
+}
+
+
+} // namespace crpropa
+
+#endif // CRPROPA_HAVE_FFTW3F
```

Comment:
You have separated the old code into logical functions, too? I like it.

Reply:
Yes. Thanks! :-)